hack/lib/golang.sh

```diff
@@ -441,6 +450,95 @@ kube::golang::outfile_for_binary() {
+# Returns 0 if the binary can be built with coverage, 0 otherwise.
+# NB: this ignores whether coverage is globally enabled or not.
+kube::golang::is_covered_binary() {
+  return $(kube::util::array_contains "$1" "${KUBE_COVERAGE_INSTRUMENTED_PACKAGES[@]}")
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler)
+# Echos the path to a dummy test used for coverage information.
+kube::golang::path_for_coverage_dummy_test() {
+	local package=$1
+  local path="${KUBE_GOPATH}/src/$package"
+  local name=$(basename "$package")
+  echo "$path/zz_autogenerated_${name}_test.go"
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler).
+# Creates a dummy unit test on disk in the source directory for the given package.
+# This unit test will invoke the package's standard entry point when run.
+kube::golang::create_coverage_dummy_test() {
+  local package="$1"
+  local name="$(basename "$package")"
+  cat <<EOF > $(kube::golang::path_for_coverage_dummy_test "$package")
+package main
+import (
+  "testing"
+  "k8s.io/kubernetes/pkg/util/coverage"
+)
+
+func TestMain(m *testing.M) {
+  // Get coverage running
+  coverage.InitCoverage("${name}")
+
+ 	// Go!
+  main()
+
+  // Make sure we actually write the profiling information to disk, if we make it here.
+  // On long-running services, or anything that calls os.Exit(), this is insufficient,
+  // so be sure to call this from inside the binary too.
+  coverage.FlushCoverage()
+}
+EOF
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler).
+# Deletes a test generated by kube::golang::create_coverage_dumy_test.
+kube::golang::delete_coverage_dummy_test() {
+  local package=$1
+  rm $(kube::golang::path_for_coverage_dummy_test "$package")
+}
+
+# Arguments: a list of kubernetes packages to build.
+# Expected variables: build_args should be set to an array of Go build arguments.
+# In addition, the standard build globals are assumed.
+#
+# Invokes Go to actually build some packages. If coverage is disabled, simply invokes
+# go install. If coverage is enabled, builds covered binaries using go test, temporarily
+# producing the required unit test files and then cleaning up after itself.
+# Non-covered binaries are then built using go install as usual.
+kube::golang::build_some_binaries() {
+  if [[ -n "${build_with_coverage:-}" ]]; then
+    local -a uncovered=()
+    for package in "$@"; do
+      if kube::golang::is_covered_binary "$package"; then
+        V=2 kube::log::info "Building $package with coverage..."
+        kube::golang::create_coverage_dummy_test "$package"
+        go test -c -o "$(kube::golang::outfile_for_binary "$package" "$platform")" \
+          -covermode count \
+          -coverpkg k8s.io/... \
+          "${build_args[@]}" \
+          -tags coverage \
+          "$package"
+        kube::golang::delete_coverage_dummy_test "$package"
```

Comment:
I'd like this to be trapped, so it can be cleaned up even if the previous command exits non-zero